tsl/src/compression/compression.c

```diff
@@ -2042,3 +2048,199 @@ update_compressed_chunk_relstats(Oid uncompressed_relid, Oid compressed_relid)
+			FormData_hypertable_compression *fd =
+				ts_hypertable_compression_get_by_pkey(hypertable_id, attname);
+
+			/*
+			 * There are 3 possible scenarios we have to consider
+			 * when dealing with columns which are part of unique
+			 * constraints.
+			 *
+			 * 1. Column is segmentby-Column
+			 * In this case we can add a single ScanKey with an
+			 * equality check for the value.
+			 * 2. Column is orderby-Column
+			 * In this we can add 2 ScanKeys with range constraints
+			 * utilizing batch metadata.
+			 * 3. Column is neither segmentby nor orderby
+			 * In this case we cannot utilize this column for
+			 * batch filtering as the values are compressed and
+			 * we have no metadata.
+			 */
+
+			if (COMPRESSIONCOL_IS_SEGMENT_BY(fd))
+			{
+				bool isnull;
+				Datum value = slot_getattr(slot, attno, &isnull);
+
+				TypeCacheEntry *tce =
+					lookup_type_cache(decompressor.in_desc->attrs[attno - 1].atttypid,
+									  TYPECACHE_EQ_OPR_FINFO);
+
+				if (!tce || !OidIsValid(tce->eq_opr_finfo.fn_oid))
+					elog(ERROR,
+						 "no equality function for type \"%s\"",
+						 format_type_be(decompressor.in_desc->attrs[attno - 1].atttypid));
+
+				if (isnull)
+				{
+					ScanKeyEntryInitialize(&scankeys[index],
+										   SK_ISNULL | SK_SEARCHNULL,
+										   attno,
+										   InvalidStrategy, /* no strategy */
+										   InvalidOid,		/* no strategy subtype */
+										   InvalidOid,		/* no collation */
+										   InvalidOid,		/* no reg proc for this */
+										   (Datum) 0);		/* constant */
+				}
+				else
+				{
+					ScanKeyEntryInitialize(&scankeys[index],
+										   0, /* flags */
+										   attno,
+										   BTEqualStrategyNumber,
+										   InvalidOid, /* No strategy subtype. */
+										   decompressor.in_desc->attrs[attno - 1].attcollation,
+										   tce->eq_opr_finfo.fn_oid,
+										   value);
+				}
+
+				index++;
+			}
+		}
+	}
+
+	*num_scankeys = index;
+	return scankeys;
+}
+
+void
+decompress_overlapping_batches(Chunk *chunk, TupleTableSlot *slot)
+{
+	Relation out_rel = table_open(chunk->table_id, RowExclusiveLock);
+
+	if (!ts_indexing_relation_has_primary_or_unique_index(out_rel))
+	{
+		/*
+		 * If there are no unique constraints there is nothing to do here.
+		 */
+		table_close(out_rel, RowExclusiveLock);
+		return;
+	}
+
+	Chunk *comp = ts_chunk_get_by_id(chunk->fd.compressed_chunk_id, true);
+	Relation in_rel = relation_open(comp->table_id, RowExclusiveLock);
+
+	RowDecompressor decompressor = build_decompressor(in_rel, out_rel);
+	Bitmapset *unique_columns = relation_unique_columns(out_rel);
+
+	int num_scankeys;
+	ScanKeyData *scankeys =
+		build_scankeys(chunk->fd.hypertable_id, decompressor, unique_columns, slot, &num_scankeys);
+
+	TableScanDesc heapScan =
+		table_beginscan(in_rel, GetTransactionSnapshot(), num_scankeys, scankeys);
+
+	for (HeapTuple compressed_tuple = heap_getnext(heapScan, ForwardScanDirection);
```

Comment:
Suspiciously low coverage for this loop, is it codecov mistake or missing tests?

Reply:
tests are WIP